### DISCUSSION TITLE
alfa-playwright: how do you check for all elements?

### DISCUSSION THREAD

**Discussion by lovebes** [Siteimprove/alfa#918] [Questions] [answered]
Hi, I'm trying to run playwright to check all elements for accessibility.

I was able to run the sample code:
```
const x = await page.$("html")
await expect(x).to.be.accessible();
```

But I'd like to check all the elements in the website launch page.

How would this be done?

Thanks in advance.

**Comment by kasperisager** [comment, marked as answer]
Hi there and thanks for reaching out! The accessibility audit is run on the entire tree and so passing the root element is enough to ensure that all elements are tested. Are you not seeing the results you're expecting?

**Reply by lovebes** [reply]
Oh I see. The result I was expecting was some sort of a result coming out from `custom-testing/scraping/test.ts` in `alfa-examples`. 

I know that one element would not be accessible.

Would this be because I need to explicitly set WCAG 2.0 somehow? I didn't know where I would set that using playwright.

Lastly - is there a way for the errors/information to show up next to the elements in "headed" browser?

**Reply by kasperisager** [reply]
You can find a complete example of using Alfa and Playwright together here: https://github.com/Siteimprove/alfa-examples/blob/main/end-to-end-testing/playwright/test/page.spec.ts. If you've configured the default rule set then all the WCAG related rules will already be running. For reference, this is the entire code:

```ts
/// <reference types="node" />
/// <reference types="mocha" />

import * as chai from "chai";
import * as playwright from "playwright";

import { Future } from "@siteimprove/alfa-future";
import { Playwright } from "@siteimprove/alfa-playwright";

import * as alfa from "@siteimprove/alfa-chai";
import rules from "@siteimprove/alfa-rules";

import { persist } from "../../../common/persist";

chai.use(
  alfa.Chai.createPlugin(
    (value: Playwright.Type) => Future.from(Playwright.toPage(value)),
    rules,
    [persist(() => "test/outcomes/page.spec.json")]
  )
);

const { expect } = chai;

describe("page.html", () => {
  let browser: playwright.Browser;
  let page: playwright.Page;

  before(async () => {
    browser = await playwright.chromium.launch();
    page = await browser.newPage();

    await page.goto(`file://${require.resolve("./fixtures/page.html")}`);
  });

  after(async () => {
    await browser.close();
  });

  it("should be accessible", async () => {
    await expect(await page.$("html")).to.be.accessible();
  });
});
```

Pay attention to the `alfa.Chai.createPlugin(...)` call which configures Alfa for use with Chai. All rules from `@siteimprove/alfa-rules` are enabled and any failed outcomes are written as EARL to `test/outcomes/page.spec.json`.

> Lastly - is there a way for the errors/information to show up next to the elements in "headed" browser?

There isn't, no. Alfa only provides data as output and leaves all reporting to external tools.

**Reply by kasperisager** [reply]
For reference, this is the `persist()` handler:

```ts
/// <reference types="node" />

import * as fs from "fs";
import * as path from "path";

import { Handler } from "@siteimprove/alfa-assert";
import { Formatter } from "@siteimprove/alfa-formatter";
import { Future } from "@siteimprove/alfa-future";
import { Mapper } from "@siteimprove/alfa-mapper";

import earl from "@siteimprove/alfa-formatter-earl";

export function persist<I, T, Q, S>(
  output: Mapper<I, string>,
  format: Formatter<I, T, Q, S> = earl()
): Handler<I, T, Q, S> {
  return (input, rules, outcomes, message) =>
    Future.from(async () => {
      const file = path.relative(process.cwd(), output(input));
      const dir = path.dirname(file);

      fs.mkdirSync(dir, { recursive: true });
      fs.writeFileSync(file, format(input, rules, outcomes) + "\n");

      return `${message}, see the full report at ${file}`;
    });
}
```

**Reply by lovebes** [reply]
Thank you so much for the detailed information! I'll play around with this code :)

**Reply by lovebes** [reply]
You know, I changed the first code you sent me with an actual website:

```
    await page.goto(`https://www.pennyjuice.com/`);
    await page.waitForSelector('.ae-icon-launcher-universalaccess', {
      waitFor: 'visible',
    })
    const x = await page.$("html")
    
    await expect(x).to.be.accessible();
```

I can see that playwright is waiting until that A18y button shows up.

However, that assertion returns a true.

If I run the same website with the SiteImprove Chrome Extension, I get a pretty sizeable list of errors.

Am I doing something wrong? Is the default rule set at the most lenient WCAG level?

**Reply by kasperisager** [reply]
I adjusted the code to the following:

```ts
/// <reference types="node" />
/// <reference types="mocha" />

import * as chai from "chai";
import * as playwright from "playwright";

import { Future } from "@siteimprove/alfa-future";
import { Playwright } from "@siteimprove/alfa-playwright";

import * as alfa from "@siteimprove/alfa-chai";
import rules from "@siteimprove/alfa-rules";

import { persist } from "../../../common/persist";

chai.use(
  alfa.Chai.createPlugin(
    (value: Playwright.Type) => Future.from(Playwright.toPage(value)),
    rules,
    [persist(() => "test/outcomes/page.spec.json")]
  )
);

const { expect } = chai;

describe("page.html", () => {
  let browser: playwright.Browser;
  let page: playwright.Page;
  let document: playwright.JSHandle;

  before(async () => {
    browser = await playwright.chromium.launch();
    page = await browser.newPage();

    await page.goto(`https://www.pennyjuice.com/`);

    document = await page.evaluateHandle(() => window.document);
  });

  after(async () => {
    await browser.close();
  });

  it("should be accessible", async () => {
    await expect(document).to.be.accessible();
  });
});
```

When run with a timeout of 30000 ms (`mocha --timeout 30000`), the assertion fails on my end with 50 failed outcomes.

**Reply by lovebes** [reply]
Thank you! I will try this out.

Hm, this is still passing for me :/

By the way, how do you run this typescript file with just `mocha`? Would that be perhaps the culprit?

I had to change it to a .js file in order to run it by `npx mocha --timeout 30000 test  ./test/chai.spec.js`

**Reply by lovebes** [reply]
I did `npx mocha --require ts-node/register --timeout 30000 test  ./test/chai.spec.ts` and got
`AssertionError: expected { Object (_events, _eventsCount, ...) } to be accessible but [object Object]`

**Reply by lovebes** [reply]
Actually I can work with the `scraping` example - it's fine. I don't want to take too much of your time. Thanks!

EDIT: one final question I'd like to ask is - how do you set the test outcome to WCAG AA or less? Is there a way to set this?

**Reply by kasperisager** [reply]
> one final question I'd like to ask is - how do you set the test outcome to WCAG AA or less? Is there a way to set this?

Each rule has a list of requirements that you can filter by:

```ts
import { Criterion, Conformance } from "@siteimprove/alfa-wcag";

const aaRules = rules.filter((rule) =>
  rule.requirements.some(
    (requirement) =>
      Criterion.isCriterion(requirement) && Conformance.isAA("2.1")(requirement)
  )
);
```

This will include all rules with WCAG 2.1 A and AA criteria as required by the AA conformance level. Do note that some criteria have changed between WCAG versions, hence the need to specify the exact WCAG version of interest. If you're interested only in WCAG 2.0 AA conformance you can change `"2.1"` to `"2.0"`.

**Reply by lovebes** [reply]
wow awesome! Thank you so much!!
